### PR TITLE
Add support for internal plugins

### PR DESCRIPTION
**Pull request by rhuss**
This PR adds the possibility for internal plugins. This works by allowing code to add to a global plugin slice `plugins.InternalPlugins`. By default this slice is empty, but for a custom assembly process (outside of this project), this could be leverage to add plugins directly in the code.

The changes are minimal and well tested, but the PR was created a bit in a hurry so apologies for not having an associated issue yet (but I will create on tomorrow).

### CODE REVIEW

File: pkg/kn/plugin/manager_test.go
Comment (mattmoor) [inline pkg/kn/plugin/manager_test.go:46]: Format Go code:
```suggestion
func (t testPlugin) Name() string                 { return strings.Join(t.parts, " ") }
func (t testPlugin) Execute(args []string) error  { return nil }
```

File: pkg/kn/plugin/manager_test.go
Comment (mattmoor) [inline pkg/kn/plugin/manager_test.go:49]: Format Go code:
```suggestion
func (t testPlugin) CommandParts() []string       { return t.parts }
func (t testPlugin) Path() string                 { return "" }
```

File: pkg/kn/plugin/manager_test.go
Comment (mattmoor) [inline pkg/kn/plugin/manager_test.go:138]: Format Go code:
```suggestion
		name  string
	}{
		{[]string{"a", "b"}, "a b"},
		{[]string{"a"}, "a"},
		{[]string{"a", "c"}, "a"},
```

File: pkg/kn/plugin/manager.go
Comment (dsimansk) [inline pkg/kn/plugin/manager.go:103]:
```suggestion
	// Try to find internal plugin first
```

**Comment by rhuss** [comment]
> Any chance to update the description or link to an issue?
> This is listed as chore. Missing the context I guess.

Ah, sorry. It's kind of an habit to use the [convention commit](https://www.conventionalcommits.org/en/v1.0.0/) spec for doing my commit message, and while creating the PR with `hub` it picked up this message as title by default.

Going to adapt the title, but there is no issue yet. Do you want me to create one or could we discuss the change on this PR ?

**Comment by rhuss** [comment]
/hold

Do not merge until we have discussed the topic on an issue (which I will create tomorrow).

**Comment by rhuss** [comment]
I'm going to rebase this PR as soon as #929 is merged, but its already ready to review.

Please see also the proposal at #928 that I have opened.

/unhold

File: pkg/kn/plugin/manager_test.go
Comment (daisy-ycguo) [inline pkg/kn/plugin/manager_test.go:120]: Besides testing a simple case with only internal plugins, I think we could add a complicated test case with both internal plugins and external plugins to list, either to enhance the test of `ListPlugins` or the test of `ListPluginsForCommandGroup` for this complicated test case.

**Comment by maximilien** [comment]
@rhuss I am not able to checkout your PR. I get the following:\

```bash

➜  client git:(master) gh pr checkout 902
From https://github.com/knative/client
 ! [rejected]          refs/pull/902/head -> pr/internal-plugins  (non-fast-forward)
exit status 1
```

Could it be a need to rebase? OR maybe I need to do something on my side...  not sure. Let me digg some more

**Comment by rhuss** [comment]
@maximilien  looks strange. Ok, let me try (but I'm typically not using `gh` but standard `git`)

**Comment by rhuss** [comment]
Hmm, I tried it with standard git commands, no issue. Maybe its a problem of `gh` ?

```
git clone https://github.com/knative/client.git
cd client
git remote add rhuss https://github.com/rhuss/knative-client.git
git fetch rhuss
git checkout -b internal-plugins rhuss/pr/internal-plugins
```

**Comment by rhuss** [comment]
I was even able to checkout with `gh`. @maximilien maybe you already have a branch `pr/internal-plugins` ? If so I sugges to remove it first with `git branch -D pr/internal-plugins` and then try again ....

File: pkg/kn/plugin/manager_test.go
Comment (rhuss) [reply pkg/kn/plugin/manager_test.go:120]: Good idea, let me add this.

**Comment by maximilien** [comment]
> git branch -D pr/internal-plugins

Worked! Thank you. Will complete my review today. Best.

File: pkg/kn/plugin/manager.go
Comment (maximilien) [inline pkg/kn/plugin/manager.go:200]: I can comment all these lines 193-200 so that function returns true always and all tests still pass...

File: pkg/kn/plugin/manager.go
Comment (maximilien) [inline pkg/kn/plugin/manager.go:188]: I can comment lines 184-188 so that no filtering happens and all tests still pass...

File: pkg/kn/plugin/manager.go
Comment (maximilien) [inline pkg/kn/plugin/manager.go:140]: I can change this line to set to `false` instead of `true` and all tests still pass...

**Comment by rhuss** [comment]
@maximilien @daisy-ycguo Thanks for the review ! Unfortunately, I wasn't able to add the remaining tests before my PTO (sorry, was just to busy ...) We can either wait until I'm back (then, I probably will have to resolve some conflicts), or we could merge now (as long as it is still mergable) and I will add the missing tests later (promised :)

See ya in 2 weeks ...

**Comment by maximilien** [comment]
Enjoy your PTO @rhuss! Talk when back. Cheers 🍻 🏖️

**Comment by rhuss** [comment]
... back in town, contiue soon on this one so that it gets included in 0.17 (due next Tuesday)

File: pkg/kn/plugin/manager.go
Comment (rhuss) [reply pkg/kn/plugin/manager.go:200]: fixed

File: pkg/kn/plugin/manager.go
Comment (rhuss) [reply pkg/kn/plugin/manager.go:140]: fixed

**Comment by rhuss** [comment]
The integration test error is not a flake but because of a change in eventing 0.17. We need to fix this before releasing kn 0.17 regardless of whether we get this PR in or not.

**Comment by rhuss** [comment]
@maximilien knative now depends on go 1.14, so you need to upgrade your go installation

**Comment by rhuss** [comment]
@dsimansk @navidshaikh @maximilien I think we are good to merge. I adressed all PR comments, and added some extra testing. Also the CI is fixed (will also fix CI issues of other PRs).

Can I have a lgtm please ? (if there are no objections)

File: CHANGELOG.adoc
Comment (navidshaikh) [inline CHANGELOG.adoc:78]: 902, can be fixed in update to changelog PR for release as well.

**Review by navidshaikh — APPROVED** [review]
/lgtm

**Pull request merged**
